circuits/src/traits.rs

```diff
@@ -0,0 +1,1193 @@
+//! Defines traits which groups types of types and translations between them
+//!
+//! We strongly type inputs to our ZK(MPC) circuits to gain circuit readability
+//! and inherit safety properties from the type checker and linters (i.e. unused
+//! witness elements)
+//!
+//! We group types of types by traits which associate with other types. For example, types
+//! allocated in an MPC circuit implement different traits than ZK circuit types do, due to
+//! their different underlying primitives
+//!
+//! At a high level the types are:
+//!     - Base types: application level types that have semantically meaningful values
+//!     - Single-prover variable types: base types allocated in a single-prover constraint system
+//!     - Single-prover commitment types: commitments to base types in a single-prover system
+//!     - MPC types: base types that have been allocated in an MPC fabric
+//!     - Multi-prover variable types: base types allocated in a multi-prover constraint system
+//!     - Multi-prover commitment types: commitments to base types in a multi-prover system
+
+use crypto::fields::{biguint_to_scalar, scalar_to_biguint};
+use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
+use itertools::Itertools;
+use mpc_bulletproof::{
+    r1cs::{
+        LinearCombination, Prover, R1CSProof, RandomizableConstraintSystem, Variable, Verifier,
+    },
+    r1cs_mpc::{MpcLinearCombination, MpcProver, MpcVariable, SharedR1CSProof},
+};
+use mpc_ristretto::{
+    authenticated_ristretto::AuthenticatedCompressedRistretto,
+    authenticated_scalar::AuthenticatedScalar, beaver::SharedValueSource,
+    mpc_scalar::scalar_to_u64, network::MpcNetwork,
+};
+use num_bigint::BigUint;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    errors::{MpcError, ProverError, VerifierError},
+    mpc::SharedFabric,
+    LinkableCommitment,
+};
+
+// ---------------
+// | Type Traits |
+// ---------------
+
+/// Implementing types are base (application level) types that define serialization to/from `Scalars`
+///
+/// Commitment, variable, MPC, etc types are implemented automatically from serialization and deserialization
+pub trait BaseType: Clone {
+    /// Convert the base type to its serialized scalar representation in the circuit
+    fn to_scalars(self) -> Vec<Scalar>;
+    /// Convert from a serialized scalar representation to the base type
+    fn from_scalars<I: Iterator<Item = Scalar>>(i: &mut I) -> Self;
```

Comment:
They are typically done in tandum, so `serialize` -> `<some ops>` -> `deserialize`, so it's generally safe to unwrap. Error handling these things makes the interfaces really messy, and colors the interfaces all the way up to the circuit. It's okay to panic too, this won't crash the relayer because the MPC is isolated into its own runtime